rust/uefi/stub/src/thin.rs

```diff
@@ -103,12 +105,57 @@ pub fn boot_linux(handle: Handle, mut system_table: SystemTable<Boot>) -> uefi::
             .expect("Failed to get file system handle");
         let mut file_system = FileSystem::new(file_system);
 
-        kernel_data = file_system
-            .read(&*config.kernel_filename)
-            .expect("Failed to read kernel file into memory");
-        initrd_data = file_system
-            .read(&*config.initrd_filename)
-            .expect("Failed to read initrd file into memory");
+        if system_table.boot_services().test_protocol::<uefi::proto::media::fs::SimpleFileSystem>(filesystem_protocol_params).is_ok() {
+            let mut file_system = system_table
+                .boot_services()
+                .get_image_file_system(handle)
+                .expect("Failed to get file system handle");
+
+            kernel_data = file_system
+                .read(&*config.kernel_filename)
+                .expect("Failed to read kernel file into memory");
+            initrd_data = file_system
+                .read(&*config.initrd_filename)
+                .expect("Failed to read initrd file into memory");
+        } else {
+            let loaded_image_protocol = system_table.boot_services().open_protocol_exclusive::<LoadedImage>(system_table.boot_services().image_handle())
+                .expect("Failed to open the loaded image protocol on the currently loaded image");
+
+            let mut base_code = system_table.boot_services().open_protocol_exclusive::<BaseCode>(loaded_image_protocol.device()).unwrap();
+
+            assert!(base_code.mode().dhcp_ack_received);
+            let dhcp_ack: &DhcpV4Packet = base_code.mode().dhcp_ack.as_ref();
+            let server_ip = dhcp_ack.bootp_si_addr;
+            let server_ip = IpAddress::new_v4(server_ip);
+
+            let kernel_filename = cstr8!("./bzImage");
+            let initrd_filename = cstr8!("./initrd");
+
+            let kfile_size = base_code
+                .tftp_get_file_size(&server_ip, kernel_filename)
+                .expect("failed to query file size");
+
+            let ifile_size = base_code
+                .tftp_get_file_size(&server_ip, initrd_filename)
+                .expect("failed to query file size");
+
+            assert!(kfile_size > 0);
+            assert!(ifile_size > 0);
+
+            kernel_data = Vec::with_capacity(kfile_size as usize);
+            kernel_data.resize(kfile_size as usize, 0);
```

Comment:
These can be `vec![0; kfile_size as usize]`. Clippy should tell you this as well.